[cmake] Internal depends download macro refactor V2

fuzzard opened the pull request:
## Description
This takes #20905 and then takes it to the conclusion i had intended.

Internal depends builds are refactored out into macros/functions to more easily template an internal dep build

## Motivation and context
Had phone/internet service out for the day, so got bored and spent some time working through this.
This would potentially be a way to migrate to cmake from tools/depends. Atm, the non cmake targets (DAV1D, FSTRCMP, UDFREAD) dont build outside of host=target, but the cmake targets can at least be built on osx for my testing purposes.

~~As always FFMPEG is untouched. If #21047 gets merged, i'll revisit these changes on it.~~
~~gtest is also not the full set of changes. Wasnt a priority, and can be done later for uniformity.~~

Ffmpeg and gtest are done now as well.

Theres a lot of commits, however i can squash a number of them into similar changes if desired.
This could supersede #20905 if desired to go all the way.

I'm ok if this gets rejected, i had a lot of time to kill today, haha. Always happy for any feedback, or opinions on changes.

## How has this been tested?
configure and build of targets where possible on osx.
minimal testing for those targets that dont build on osx on a ubuntu vm

## What is the effect on users?
n/a

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[tools/depends] rename libfmt to fmt

rename libfmt to just fmt. This brings some consistency to naming in other parts
of kodi cmake (eg target fmt)

Commit:
[tools/depends] rename libfstrcmp to fstrcmp

rename libfstrcmp to just fstrcmp. This brings some consistency to naming in other parts
of kodi cmake (eg target fstrcmp)

Commit:
[tools/depends] rename libspdlog to spdlog

rename libspdlog to just spdlog. This brings some consistency to naming in other parts
of kodi cmake (eg target spdlog)

Commit:
[cmake] create macro to handle module externalproject_add calls for internal deps

refactor out usage of externalproject_add to a macro for internal dep building

Commit:
[cmake] FindCrossGUID use BUILD_DEP_TARGET macro

Commit:
[cmake] FindDav1d use BUILD_DEP_TARGET macro

Commit:
[cmake] FindFlatbuffers use BUILD_DEP_TARGET macro

Commit:
[cmake] FindFMT use BUILD_DEP_TARGET macro

Commit:
[cmake] FindFstrcmp use BUILD_DEP_TARGET macro

Commit:
[cmake] FindRapidJSON use BUILD_DEP_TARGET macro

Commit:
[cmake] FindSpdlog use BUILD_DEP_TARGET macro

Commit:
[cmake] FindUdfread use BUILD_DEP_TARGET macro

Commit:
[tools/depends] introduce BYPRODUCT variable in VERSION files

byproduct tracking is used by both tools/depends and project cmake build system

Commit:
[cmake] ModuleHelper read BYPRODUCT from VERSION files

Commit:
[cmake] SETUP_BUILD_VARS create MODULE_LIBRARY/INCLUDE_DIR/VER variables

Commit:
[cmake] Find CrossGUID utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] FindDav1d utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] FindFlatbuffers utilise automated variables

Use automated VER variable

Commit:
[cmake] FindFMT utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] FindFstrcmp utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] FindRapidJSON utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] FindSpdlog utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] FindUdfread utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] BUILD_DEP_TARGET move include and add docinfo

Commit:
[tools/depends] rename libudfread to udfread

rename libudfread to just udfread. This brings some consistency to naming in other parts
of kodi cmake (eg target udfread)

Commit:
[cmake] FindDav1d do find_program tests for meson/ninja for internal build

Commit:
[tools/depends] rename googletest to gtest

rename googletest to just gtest. This brings some consistency to naming in other parts
of kodi cmake (eg target gtest)

Commit:
[cmake] FindGtest utilise automated variables

Use automated BYPRODUCT, INCLUDE_DIRS, LIBRARY, VER variables

Commit:
[cmake] FindFFMPEG utilise automated variables and macros

Use SETUP_BUILD_VARS and BUILD_DEP_TARGET macros
Use automated INCLUDE_DIRS, LIBRARY, VER variables

fuzzard commented:
jenkins build this please

wsnipex commented on cmake/scripts/common/ModuleHelpers.cmake:
```diff
@@ -66,4 +72,75 @@ macro(SETUP_BUILD_VARS)
+    set(BUILD_BYPRODUCTS ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/${${MODULE}_BYPRODUCT})
+  else()
+    set(BUILD_BYPRODUCTS ${BUILD_BYPRODUCTS})
+  endif()
```
you could invert the if for consistency and readablility

fuzzard replied:
Will do

fuzzard replied:
Have fixed that.
Also changed to BUILD_BYPRODUCTS is an optional. In the case of ffmpeg, we dont set a byproduct (it has more than one essentially), so checks have been put in to handle no byproduct, and to not set the byproduct vars if they arent found in the VERSION file.
This is also the case for MODULE_LIBRARY, if there isnt a byproduct, dont generate that LIBRARY var automatically. This just means the dev has to set the MODULE_LIBRARY var if its required. in the case of ffmpeg, we dont use FFMPEG_LIBRARY, so no need for it to be manually set.

Commit:
[cmake] BUILD_DEP_TARGET append Toolchain when CMAKE_ARGS passed

If a toolchain file is provided, pass it along to any internal dep cmake based
build systems.
